java/ql/src/Stubs/Stubs.qll

```diff
@@ -0,0 +1,440 @@
+    then result = "{}"
+    else result = "{ return " + stubDefaultValue(c.getReturnType()) + "; }"
+}
+
+private string stubDefaultValue(Type t) {
+  if t instanceof RefType
+  then result = "null"
+  else
+    if t instanceof CharacterType
+    then result = "'0'"
+    else
+      if t instanceof BooleanType
+      then result = "false"
+      else
+        if t instanceof NumericType
+        then result = "0"
+        else result = "<error>"
+}
+
+private string stubParameters(Callable c) {
+  result =
+    concat(int i, Parameter param |
+      param = c.getParameter(i)
+    |
+      stubParameter(param), ", " order by i
+    )
+}
+
+private string stubParameter(Parameter p) {
+  exists(Type t, string suff | result = stubTypeName(t) + suff + " " + p.getName() |
+    if p.isVarargs()
+    then (
+      t = p.getType().(Array).getElementType() and
+      suff = "..."
+    ) else (
+      t = p.getType() and suff = ""
+    )
+  )
+}
+
+private string stubEnumConstants(RefType t) {
+  if t instanceof EnumType
+  then
+    exists(EnumType et | et = t |
+      result =
+        "    " + concat(EnumConstant c | c = et.getAnEnumConstant() | c.getName(), ", ") + ";\n"
+    )
+  else result = ""
+}
+
+// Holds if the member is to be excluded from stubMember
+private predicate excludedMember(Member m) {
+  m instanceof EnumConstant
+  or
+  exists(Method c | m = c |
+    c.getDeclaringType() instanceof EnumType and
+    m.hasName(["values", "valueOf"]) and
+    m.isStatic()
+  )
+}
+
+private string stubMember(Member m) {
+  if excludedMember(m)
+  then result = ""
+  else (
+    exists(Method c | m = c |
+      result =
+        "    " + stubModifiers(c) + stubGenericMethodParams(c) + stubTypeName(c.getReturnType()) +
+          " " + c.getName() + "(" + stubParameters(c) + ")" + stubImplementation(c) + "\n"
+    )
+    or
+    exists(Constructor c | m = c |
+      result =
+        "    " + stubModifiers(m) + c.getName() + "(" + stubParameters(c) + ")" +
+          stubImplementation(c) + "\n"
+    )
+    or
+    exists(Field f | f = m |
+      result =
+        "    " + stubModifiers(m) + stubTypeName(f.getType()) + " " + f.getName() + " = " +
+          stubDefaultValue(f.getType()) + ";\n"
+    )
+    or
+    exists(NestedType nt | nt = m | result = indent(nt.(GeneratedType).getStub()))
+  )
+}
+
+bindingset[s]
+private string indent(string s) { result = "    " + s.replaceAll("\n", "\n    ") + "\n" }
+
+// If a class's superclass doesn't have a no-arg constructor, then it won't compile when its constructor's bodies are stubbed
+// So we synthesise no-arg constructors for each generated type that doesn't have one.
+private string stubFakeConstructor(RefType t) {
+  if not t instanceof Class
+  then result = ""
+  else
+    exists(string mod |
+      if t instanceof EnumType then mod = "    private " else mod = "    protected "
+    |
+      if hasNoArgConstructor(t) then result = "" else result = mod + t.getName() + "() {}\n"
```

Comment:
Won't this generate a duplicate no-arg private constructor for an enum type that already has one? (If any existing one is being discarded elsewhere, note this with a comment)